src/DPORDriver.cpp

```diff
@@ -192,17 +195,209 @@ Trace *DPORDriver::run_once(TraceBuilder &TB, bool &assume_blocked) const{
   return t;
 }
 
+void DPORDriver::print_progress(uint64_t computation_count, long double estimate, Result &res) {
+  if(computation_count % 100 == 0){
+    llvm::dbgs() << ESC_char << "[K" // Erase the line
+                 << "Traces: " << res.trace_count;
+    if(res.sleepset_blocked_trace_count)
+      llvm::dbgs() << ", " << res.sleepset_blocked_trace_count << " ssb";
+    if(res.assume_blocked_trace_count)
+      llvm::dbgs() << ", " << res.assume_blocked_trace_count << " ab";
+    if(conf.print_progress_estimate){
+      std::stringstream ss;
+      ss << std::setprecision(LDBL_DIG) << estimate;
+      llvm::dbgs() << " ("
+                   << int(100.0*(long double)(computation_count+1)/estimate)
+                   << "% of total estimate: "
+                   << ss.str() << ")";
+    }
+    llvm::dbgs() << "\r"; // Move cursor to start of line
+  }
+}
+
+bool DPORDriver::handle_trace(TraceBuilder *TB, Trace *t, uint64_t *computation_count, Result &res, bool assume_blocked) {
+  bool t_used = false;
+  if(t && conf.debug_collect_all_traces){
+    res.all_traces.push_back(t);
+    t_used = true;
+  }
+  if(!TB->sleepset_is_empty()) {
```

Comment:
There's a race on `TB` here. Since the threadpool queue might have been non-empty, the worker thread might have moved on to process a new work item.

Reply:
RFSCTraceBuilder always returns true here which is why we let it race since it never actually affects anything. If the `sleepset_is_empty()` would be changed in the future that will cause a problem, so we should change it as that would also remove the tracebuilder entirely from the required argument parameters.

Reply:
I see. Still, this code is a liability, especially so without a comment.

Reply:
While we are in agreement of that it should change, there is a comment upon the parallel version since the rest are not parallelized:
```
// handle_trace requires a TB but with RFSC this is a constant value, so it does not matter which TB it is.
if (handle_trace(TBs[0], t, &computation_count, res, assume_blocked)) {
    ...
}
```